File: src/app/services/part.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Part } from '../models/part.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class PartService {
  private apiUrl = `${environment.apiUrl}/parts`;

  constructor(private http: HttpClient) {}

  validateAndUpdatePartDescription(partDescription: string): Observable<any> {
    const url = `${this.apiUrl}/validateAndUpdatePartDescription`;
    return this.http.post(url, { partDescription });
  }

  validateNextItem(currentItem: any): Observable<any> {
    const url = `${this.apiUrl}/validateNextItem`;
    return this.http.post(url, { currentItem });
  }

  fetchPartNumbers(globalParameter: number, unitId: string, groupId: string, lineId: string): Observable<any> {
    const url = `${this.apiUrl}/fetchPartNumbers`;
    return this.http.post(url, { globalParameter, unitId, groupId, lineId });
  }

  validatePartNumber(partNumber: string, globalParameter: number, unitId: string, groupId: string, lineId: string): Observable<any> {
    const url = `${this.apiUrl}/validatePartNumber`;
    return this.http.post(url, { partNumber, globalParameter, unitId, groupId, lineId });
  }

  checkPartExists(UNIT_ID: string, GROUP_ID: string, LINE_ID: string, PARTNO: string): Observable<any> {
    const url = `${this.apiUrl}/checkPartExists`;
    return this.http.post(url, { UNIT_ID, GROUP_ID, LINE_ID, PARTNO });
  }

  savePart(UNIT_ID: string, GROUP_ID: string, LINE_ID: string, PARTNO: string, PART_DESC: string, PART_STATUS: string): Observable<any> {
    const url = `${this.apiUrl}/savePart`;
    return this.http.post(url, { UNIT_ID, GROUP_ID, LINE_ID, PARTNO, PART_DESC, PART_STATUS });
  }

  updatePart(UNIT_ID: string, GROUP_ID: string, LINE_ID: string, PARTNO: string, PART_DESC: string, PART_STATUS: string, PART_ID: string): Observable<any> {
    const url = `${this.apiUrl}/updatePart`;
    return this.http.post(url, { UNIT_ID, GROUP_ID, LINE_ID, PARTNO, PART_DESC, PART_STATUS, PART_ID });
  }
}

File: src/main/java/com/example/tmm022_fmb/service/PartService.java
package com.example.tmm022_fmb.service;

import com.example.tmm022_fmb.repository.PartRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

@Service
public class PartService {

    @Autowired
    private PartRepository partRepository;

    public String validateAndUpdatePartDescription(String partDescription) {
        // Business logic to validate and update Part Description
        if (partDescription == null || partDescription.isEmpty()) {
            return "Part Description cannot be empty";
        }
        // Assuming partRepository.findByDescription is a method to find part by description
        if (partRepository.findByDescription(partDescription) != null) {
            return "Part Description already exists";
        }
        // Update logic here
        return "Part Description updated successfully";
    }

    public String validateNextItem(Map<String, Object> currentItem) {
        // Business logic to validate the next item
        if (currentItem.get("UNIT_ID") == null || currentItem.get("UNIT_NAME") == null) {
            return "Unit ID and Unit Name should not be null";
        }
        if (currentItem.get("GROUP_ID") == null || currentItem.get("GROUP_NAME") == null) {
            return "Group ID and Group Name should not be null";
        }
        if (currentItem.get("LINE_ID") == null || currentItem.get("LINE_DESC") == null) {
            return "Line ID and Line Name should not be null";
        }
        // Assuming globalParameter is a field in currentItem
        int globalParameter = (int) currentItem.get("globalParameter");
        if (globalParameter == 0) {
            if (currentItem.get("PARTNO") == null) {
                return "Part No and Part Description should not be null";
            }
            // Set PART_STATUS to 'A'
            currentItem.put("PART_STATUS", "A");
        } else if (globalParameter == 1) {
            if (currentItem.get("PART_ID") == null) {
                return "Kindly Choose data from LOV before changing Description";
            }
            if (currentItem.get("PARTNO") == null || currentItem.get("PART_DESC") == null) {
                return "Part No and Part Description should not be null";
            }
            // Set PART_STATUS to 'A'
            currentItem.put("PART_STATUS", "A");
        }
        return "Validation successful";
    }

    public String validatePartStatus(String partStatus) {
        if (partStatus == null || partStatus.isEmpty()) {
            return "PART_STATUS cannot be null";
        }
        return "PART_STATUS is valid";
    }

    public List<String> validateRequiredFields(Map<String, Object> fields) {
        List<String> errorMessages = new ArrayList<>();
        for (Map.Entry<String, Object> entry : fields.entrySet()) {
            if (entry.getValue() == null || entry.getValue().toString().isEmpty()) {
                errorMessages.add(entry.getKey() + " cannot be null");
            }
        }
        return errorMessages;
    }

    public String validateUnitIdAndName(String unitId, String unitName, int globalParameter) {
        if (globalParameter == 0) {
            if (partRepository.findByUnitIdAndUnitName(unitId, unitName) == null) {
                return "Unit ID and Unit Name do not exist in MES_UNIT_MASTER";
            }
        } else if (globalParameter == 1) {
            if (partRepository.findByUnitIdAndUnitNameInBothTables(unitId, unitName) == null) {
                return "Unit ID and Unit Name do not exist in both MES_UNIT_MASTER and HPM_PART_MASTER";
            }
        }
        return "Unit ID and Unit Name are valid";
    }

    public List<String> getLineIdList() {
        return partRepository.getLineIdList();
    }

    public String validateLineId(String lineId, String lineDesc, String unitId, String groupId) {
        if (partRepository.validateLineId(lineId, lineDesc, unitId, groupId) == null) {
            return "Line ID and Line Description do not exist";
        }
        return "Line ID and Line Description are valid";
    }

    public boolean checkPartExists(String unitId, String groupId, String lineId, String partNo) {
        return partRepository.existsByUnitIdAndGroupIdAndLineIdAndPartNo(unitId, groupId, lineId, partNo);
    }

    public String savePart(String unitId, String groupId, String lineId, String partNo, String partDesc, String partStatus) {
        if (checkPartExists(unitId, groupId, lineId, partNo)) {
            return "Part already exists";
        }
        partRepository.save(new Part(unitId, groupId, lineId, partNo, partDesc, partStatus));
        return "Part saved successfully";
    }

    public String updatePart(String unitId, String groupId, String lineId, String partNo, String partDesc, String partStatus, String partId) {
        if (!checkPartExists(unitId, groupId, lineId, partNo)) {
            return "Part does not exist";
        }
        Part part = new Part(unitId, groupId, lineId, partNo, partDesc, partStatus);
        part.setPartId(partId);
        partRepository.save(part);
        return "Part updated successfully";
    }
}